From 563a19a683d6e0e7b1218e4558a0fb75333b5b34 Mon Sep 17 00:00:00 2001
From: Ondrej Holy <oholy@redhat.com>
Date: Tue, 28 Apr 2026 04:14:23 +0000
Subject: [PATCH] [allocations] fix growth of preallocated buffers

Partial backport of commit 118afc0b954ba9d5632b7836ad24e454555ed113.

Only the `Stream_EnsureCapacity` fix is included. Other changes from
the upstream commit (`sizeof(WCHAR)` replacements, collection growth
patterns, environment block handling) are unrelated to CVE-2026-27951.

Made-with: Cursor
---
 winpr/libwinpr/utils/stream.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/winpr/libwinpr/utils/stream.c b/winpr/libwinpr/utils/stream.c
index cc119c7..cb145b0 100644
--- a/winpr/libwinpr/utils/stream.c
+++ b/winpr/libwinpr/utils/stream.c
@@ -40,8 +40,10 @@ BOOL Stream_EnsureCapacity(wStream* s, size_t size)
 
 		do
 		{
-			new_capacity *= 2;
-		} while (new_capacity < size);
+			if (new_capacity > SIZE_MAX - 128ull)
+				return FALSE;
+			new_capacity += 128ull;
+		} while (new_capacity <= size);
 
 		position = Stream_GetPosition(s);
 
-- 
2.53.0

